HCIDOCS-75: Add IPI installation workflow information to our document… (#94545)
openshift/openshift-docs: modules/con_ipi-installation-workflow.adoc

// Module included in the following assemblies:
//
// * installing/installing_bare_metal/ipi/ipi-install-prerequisites.adoc

:_mod-docs-content-type: CONCEPT

[id="installation-workflow_{context}"]
= Installation workflow

> The installation program supports interactive mode, but it is recommended to prepare an `install-config.yaml` file in advance, containing the provisioning details for all of the bare-metal hosts.

Author: Updated as suggested.

> The installation program supports interactive mode, but Red Hat recommends preparing an `install-config.yaml` file in advance, containing the provisioning details for all of the bare-metal hosts.

Author: Updated as suggested.

> Add the relevant cluster details to the `install-config.yaml` file. The installation program ships the `install-config.yaml` file. The administrator generates the manifests and verifies all prerequisites.

Reviewer: ```suggestion
Add the relevant cluster details to the `install-config.yaml` file. The installation program ships the `install-config.yaml` file. The administrator generates the manifests and verifies all prerequisites.
```

I'm not sure what "ships" means in this context. Is there a more descriptive or accurate way to phrase this?

Author: Updated as suggested.

The installation program enrolls all nodes in the cluster and starts the bootstrap VM. Metal platform components are started as `systemd` services, which have the following containers:

> * Ironic-dnsmasq: DHCP server responsible for handing over the IP addresses to the provisioning interface of various nodes on the provisioning network. Ironic-dnsmasq is only enabled when the `provisioningNetwork` is enabled.

Reviewer 2: "The DHCP server". ISG. 

The second sentence is an example where passive voice is optional, because you are focusing more on the object than the subject. You can also say, "Ironic-dnsmasq is only enabled when you deploy an {product-title} cluster with a provisioning network, or "When you deploy a cluster with a provisioning network, the installation program enables Ironic-dnsmasq." These are just a few examples of where knowing the subject of the action leads to stronger and clearer sentences.

Author: Updated as suggested.

* Ironic-httpd: http server used to ship the images to the nodes.
* Image-customization
* Ironic
* Ironic-inspector (available in {openshift} 4.16 and earlier)
* Ironic-ramdisk-logs
* Extract-machine-os

> * Provisioning-interface: Service that configures networking when the provisioning interface is enabled.

Author: Updated as suggested.

> * Metal3-baremetal-operator: Operator that provides a higher-level API for Ironic and replaces `terraform-provider-ironic` (available in {openshift} 4.16 and later)

Author: Updated as suggested.

> * Metal3-baremetal-operator: The operator that provides a higher-level API for Ironic and replaces `terraform-provider-ironic` (available in {openshift} 4.16 and later)

Reviewer: ```suggestion
* Metal3-baremetal-operator: The operator that provides a higher-level API for Ironic and replaces `terraform-provider-ironic` (available in {product-title} 4.16 and later)
```

Author: Updated as suggested.

> The nodes enter the validation phase. The nodes move to a “manageable” state after Ironic validates the credentials to access the Baseboard Management Controller (BMC) for each node.

Reviewer 2: Consider "...phase, where each node moves..." and omit "for each node" at the end.

> Once in the “manageable” state, the “inspection” phase starts. The "inspection" phase ensures that the hardware meets the minimum requirements needed for a successful deployment of {ocp}.

Reviewer: ```suggestion
Once in the “manageable” state, the “inspection” phase starts. The "inspection" phase ensures that the hardware meets the minimum requirements needed for a successful deployment of {product-title}.
```

> The `install-config.yaml` file details the provisioning network. The installation program on the bootstrap VM will try to push a live image to every node with the Ironic Python Agent (IPA) loaded in the RAM. When using virtual media, the live image connects directly to the BMC.

Reviewer 2: I don't think you need to say "in the RAM" here. Check on that last clause with QE. I would think the BMC connects with the live image, since the image is just an ISO.

Author: Updated as suggested. Will check with QE on the second point.

Author: @sgoveas, could you please confirm if the last clause on line 30 is correct?

> The `install-config.yaml` file details the provisioning network. The installation program on the bootstrap VM will try to push a live image to every node with the Ironic Python Agent (IPA) loaded. When using virtual media, the live image connects directly to the BMC.

Reviewer 3: ```suggestion
The `install-config.yaml` file details the provisioning network. The installation program on the bootstrap VM will try to push a live image via PXE to every node with the Ironic Python Agent (IPA) loaded. When using virtual media, it connects directly to each node's BMC to virtually attach the image.
```

When using PXE boot, all nodes reboot to start the process:

* The `ironic-dnsmasq` service running on the bootstrap VM provides the IP address of the node and the TFTP boot server.

> * The first boot software loads the root file system over HTTP.

Reviewer 2: It seems like "first-boot" should be hyphenated in this case.

* The `ironic` service on the bootstrap VM receives the hardware information from each node.

The nodes then enter the “cleaning” state, where each node must clean all the disks before continuing with the configuration.

Once the “cleaning” state finishes, the nodes enter the “available” state and the installation program moves the nodes to the “deploying” state.

> IPA runs the coreos-installer to install the RHCOS image on the disk defined by `rootDeviceHints` in the `install-config.yaml` file. The node boots into the new RHCOS.

Reviewer 2: Check formatting requirements for coreos-installer. That seems like a CLI name that might be a string literal in back ticks. I believe there is a variable for RHCOS.

Maybe "The node boots by using RHCOS"?

Author: Updated as suggested.

> Once the control plane nodes are configured, control moves to the control plane nodes and the bootstrap is removed.

Author: Updated as suggested.

The baremetal-operator continues the deployment of the workers, storage and infra nodes. Once the installation completes, the nodes move to the "active" state.

> The administrator performs the postinstallation checks and proceeds with Day 2 tasks.

Author: Updated as suggested.

> The baremetal-operator continues the deployment of the workers, storage and infra nodes. Once the installation completes, the nodes move to the "active" state.

Reviewer: ```suggestion
The baremetal-operator continues the deployment of the workers, storage, and infra nodes. Once the installation completes, the nodes move to the "active" state.
```

// Module included in the following assemblies:
//
// * installing/installing_bare_metal/ipi/ipi-install-prerequisites.adoc

:_mod-docs-content-type: CONCEPT

[id="installation-workflow_{context}"]
= Installation workflow

The installation program supports interactive mode, but it is recommended to prepare an `install-config.yaml` file in advance, containing the provisioning details for all of the bare-metal hosts.

The administrator adds the relevant cluster details to the `install-config.yaml` file. The `install-config.yaml` file is shipped to the provisioning host. The administrator generates the manifests and verifies all prerequisites.

The installation program enrolls all nodes in the cluster and starts the bootstrap VM. The bootstrap VM will start the ironic pod, which has the following containers:

* dnsmasq: DHCP server responsible for handing over the IP addresses to the provisioning interface of various nodes on the provisioning network.

> * httpd: http server used to ship the images to the nodes.

Author: Updated as suggested.

> * Cluster-bootstrap

Author: Updated as suggested.

* Image-customization
* Ironic
* Ironic-inspector
* Ironic-ramdisk-logs
* Coreos-downloader

The nodes enter the validation phase. The nodes move to a “manageable” state once all of the following validations succeed:

* Ironic validates that the nodes are accessible.

> * The machine network validates the credentials to access iLO for each node.

Author: Updated as suggested.

* The machine network validates access to the iLO from the primary nodes. Primary nodes need to be able to access the iLO interface of every node in the cluster (worker nodes, storage nodes and infra nodes).

Once in the “manageable” state, the “inspection” phase starts. The inspection phase ensures that the hardware meets the minimum requirements needed for a successful deployment of {ocp}.

If the `install-config.yaml` file has details for the provisioning network and PXE boot is enabled, the installation program on the bootstrap VM will try to push a live image to every node with the Ironic Python Agent (IPA) that will be loaded in the RAM.

All nodes reboot to start the PXE process:

* DHCP provides the node's IP address and the IP address of the TFTP boot server. These will be provided by the `dnsmasq` container running on the bootstrap VM.

> * Http loads the `rootfs` into the host.

Author: Updated as suggested.

> * The ironic-inspector container on the bootstrap VM receives the hardware information from each node.

Author: Updated as suggested.

The nodes then enter the “cleaning” state, where each node must clean all the disks before further configurations can be made.

Once the “cleaning” state finishes, the nodes enter the “available” state and the installation program moves the nodes to the “deploying” state.

IPA runs the coreos-installer to install the RHCOS image on the disk defined by `rootDeviceHints` in the `install-config.yaml` file. The node boots into the new RHCOS.

> Once the primary nodes are configured, control moves to the primary nodes and the bootstrap is removed.

Author: No, it should probably be control, plane nodes. Updated as suggested.
